CMakeLists.txt

```diff
@@ -799,6 +800,28 @@ add_custom_target(${DRIVER_VALIDATION_H_TARGET} DEPENDS ${DRV_VALIDATION})
 include(${ZEPHYR_BASE}/cmake/kobj.cmake)
 gen_kobj(KOBJ_INCLUDE_PATH)
 
+# Generate sections for kernel device subsystems
+set(
+  DEVICE_API_LD_SECTIONS
+  ${CMAKE_CURRENT_BINARY_DIR}/include/generated/device-api-sections.ld
+  )
+
+add_custom_command(
+  OUTPUT ${DEVICE_API_LD_SECTIONS}
+  COMMAND
+    ${PYTHON_EXECUTABLE}
+    ${ZEPHYR_BASE}/scripts/build/gen_iter_sections.py
+    -i ${struct_tags_json}
+    -t __subsystem
+    -o ${DEVICE_API_LD_SECTIONS}
```

Comment:
This unconditionally creates an ld file, however that ld file is not / cannot be used in following cases:
- `CONFIG_CMAKE_LINKER_GENERATOR=y` when using ld as a linker.
- ARM Compiler 6, which uses scatter files, ref: https://docs.zephyrproject.org/latest/develop/toolchains/arm_compiler.html.

I'm much in favor of anything which can help us improve code / type safety, but we must ensure that current functionality is still intact / migration to a new coding style does not make supporting of other toolchains unneccesary harder.

What are the plans to ensure existing functionality / support for armclang is kept when migrating other drivers ?

Reply:
See https://github.com/zephyrproject-rtos/zephyr/pull/71773#issuecomment-2107113198 but it should be supported with the manual entries of `zephyr_iterable_sections` for each subsystem that is converted.

We should add `linker_generator` test cases together with the conversion PRs.

Reply:
Example test case for ADC: https://github.com/zephyrproject-rtos/zephyr/pull/72292/commits/7857c58ae1768653a0d9cdfe7e45540678cc4c72

Reply:
@tejlmand would a CMake helper funtion/macro like `zephyr_api_section` be desired? See https://github.com/zephyrproject-rtos/example-application/pull/65#discussion_r1598297118

Reply:
thanks for the extra pieces of info.

We have `struct_tags.json` generated build time, we have the linker script generated build time.

I think we should aim for using this information to automatically also create the iterable sections for the linker generator, else the risk is too high that developers will forget to add `zephyr_iterable_section(...)`

Reply:
That would even be better, however I didn't see an easy implementation...